src/mapgen/mapgen_valleys.cpp

```diff
@@ -270,25 +280,31 @@ void MapgenValleys::makeChunk(BlockMakeData *data)
 
 int MapgenValleys::getSpawnLevelAtPoint(v2s16 p)
 {
-	// Check if in a river channel
-	float n_rivers = NoisePerlin2D(&noise_rivers->np, p.X, p.Y, seed);
-	if (std::fabs(n_rivers) <= river_size_factor)
-		// Unsuitable spawn point
-		return MAX_MAP_GENERATION_LIMIT;
-
 	float n_slope          = NoisePerlin2D(&noise_inter_valley_slope->np, p.X, p.Y, seed);
 	float n_terrain_height = NoisePerlin2D(&noise_terrain_height->np, p.X, p.Y, seed);
 	float n_valley         = NoisePerlin2D(&noise_valley_depth->np, p.X, p.Y, seed);
 	float n_valley_profile = NoisePerlin2D(&noise_valley_profile->np, p.X, p.Y, seed);
 
 	float valley_d = n_valley * n_valley;
 	float base = n_terrain_height + valley_d;
-	float river = std::fabs(n_rivers) - river_size_factor;
-	float tv = std::fmax(river / n_valley_profile, 0.0f);
-	float valley_h = valley_d * (1.0f - std::exp(-tv * tv));
-	float surface_y = base + valley_h;
-	float slope = n_slope * valley_h;
+	float surface_y = base;
 	float river_y = base - 1.0f;
+	float slope = 0.0f;
+
+	if (spflags & MGVALLEYS_CANYONS) {
+		float n_rivers = NoisePerlin3D(&noise_rivers->np, p.X, std::floor(base), p.Y, seed);
```

Comment:
I'm not sure this is ideal. Even with 'canyons' enabled, this is meant to keep spawn away from the width of the river channel as if it were 2D noise.
I would prefer this uses the existing code for both cases for simplicity. In the 3D case it's fine for it to just roughly keep spawn away from canyon areas.